airmail_indexer/src/osm_pbf.rs

```diff
@@ -0,0 +1,182 @@
+
+        debug!("Parsing POIs");
+
+        // Extract points of interest. Par_map_reduce is used to paralleliase the extraction of
+        // POIs from the OSM PBF file.
+        let pois = ElementReader::from_path(&self.pbf_path)?.par_map_reduce(
+            |element| match element {
+                // A dense node is a POI (with tags), and has the location embedded in the data.
+                Element::DenseNode(dn) => {
+                    let tags = dn
+                        .tags()
+                        .collect::<HashMap<_, _>>();
+
+                    if let Some(interesting_poi) =
+                        OsmPoi::new(tags, (dn.lat(), dn.lon())).and_then(OsmPoi::index_poi)
+                    {
+                        count_dense_nodes.fetch_add(1, Ordering::Relaxed);
+                        self.sender.send(interesting_poi).expect("sender failed");
+                        1
+                    } else {
+                        0
+                    }
+                }
+
+                // A way is a polyline or polygon, like a road.
+                Element::Way(way) => {
+                    // Attempt to get the location from the way from the underlying way data,
+                    // this requires the way is stored with the option LocationsOnWays enabled.
+                    let mut location = way
+                        .node_locations()
+                        .next()
+                        .map(|location| (location.lat(), location.lon()));
+
+                    // If the location is not present in the way data, attempt to get the location
+                    // from the node_map we previously built.
+                    if location.is_none() {
+                        location = way
+                            .refs()
+                            .next()
+                            .and_then(|node_id| {
+                                self.indexer_cache.query_node_location(node_id).ok()
+                            })
+                            .flatten();
+                    }
+
+                    if let Some(location) = location {
+                        let tags = way
+                            .tags()
+                            .collect::<HashMap<_, _>>();
+                        if let Some(interesting_poi) =
+                            OsmPoi::new(tags, location).and_then(OsmPoi::index_poi)
+                        {
+                            count_ways.fetch_add(1, Ordering::Relaxed);
+                            self.sender.send(interesting_poi).expect("sender failed");
+                            1
+                        } else {
+                            0
+                        }
+                    } else {
+                        0
+                    }
+                }
+
+                // Ignored
+                Element::Relation(_) |
+
+                // Seems to be unused, as dense node is used instead
+                Element::Node(_) => 0,
```

Comment:
I would prefer we handle this identically to dense nodes. I don't think there's any semantic difference defined between these two other than the way they're stored so there could be valid osm.pbf's out there that would only be partially indexed with this code.

Reply:
Yep, good suggestion. Originally, I had both and only removed it as clippy complained the method was too long, however, after running the planet import yesterday, I realised the nodes are needed to resolve the ways. Just double checking this at the moment.